designs/priorityqueue.md

```diff
@@ -0,0 +1,100 @@
+Priority Queue
+===================
+
+This document describes the motivation behind implementing a priority queue
+in controller-runtime and its design details.
+
+## Motivation
+
+1. Controllers reconcile all objects during startup to account for changes in
+   the reconciliation logic. Some controllers also periodically re-reconcile
+   everything to account for out of band changes they do not get notified for,
+   this is for example common for controllers managing cloud resources. In both
+   these cases, the reconciliation of new or changed objects gets delayed,
+   resulting in poor user experience. [Example][0]
+2. There may be application-specific reason why some events are more important
+   than others, [Example][1]
+
+## Proposed changes
+
+Implement a priority queue in controller-runtime that exposes the following
+interface:
+
+```go
+type PriorityQueue[T comparable] interface {
+    // AddWithOpts adds one or more items to the workqueue. Items
+    // in the workqueue are de-duplicated, so there will only ever
+    // be one entry for a given key.
+    // Adding an item that is already there may update its wait
+    // period to the lowest of existing and new wait period or
+    // its priority to the highest of existing and new priority.
+    AddWithOpts(o AddOpts, items ...T)
```

Comment:
Do we have to be able to enqueue multiple items at once? Otherwise (item, opts) would be more consistent with the embedded interface (no strong opinion from my side)

Reply:
Yeah we do have this use-case and currently do it by filtering before adding items to the workqueue: https://github.com/kubernetes-sigs/controller-runtime/blob/e3347b5405cdd0da5bff527af3d406117938ba6b/pkg/handler/enqueue_mapped.go#L93-L95

Reply:
Sounds good